crates/sui-framework/sources/governance/voting_power.move

```diff
@@ -0,0 +1,256 @@
+    use sui::validator::{Self, Validator};
+    use std::vector;
+
+    /// Set total_voting_power as 10_000 by convention. Individual voting powers can be interpreted
+    /// as easily understandable basis points (e.g., voting_power: 100 = 1%, voting_power: 1 = 0.01%) rather than
+    /// opaque quantities whose meaning changes from epoch to epoch as the total amount staked shifts.
+    /// Fixing the total voting power allows clients to hardcode the quorum threshold and total_voting power rather
+    /// than recomputing these.
+    // TODO: we can go bigger if more precision is needed (e.g., will validators have <0.01% voting power given min stake requirements)?,
+    // but I think using a power of 10 is a useful convention.
+    const TOTAL_VOTING_POWER: u64 = 10_000;
+
+    /// Quorum threshold for our fixed voting power--any message signed by this much voting power can be trusted
+    /// up to BFT assumotions
+    const QUORUM_THRESHOLD: u64 = 6_667;
+
+    /// Cap voting power of an individual validator at 10%.
+    // TODO: determine what this should be
+    const MAX_VOTING_POWER: u64 = 1_000;
+
+    /// We should never observe this, modulo bugs
+    const EInternalInvariantViolation: u64 = 0;
+
+    /// Convert each validator's stake to a voting power normalized w.r.t `TOTAL_VOTING_POWER`,
+    /// and update `active_validators` accordingly, and attempt cap each validator's voting power at `MAX_VOTING_POWER`.
+    /// Capping is handled by redistributing the voting power "taken away from" capped validators *proportionally* among validators with voting
+    /// power less than the max.
+    /// Similarly, "leftover" voting power due to rounding error is distributed *equally* among validators with voting power less than the max (if possible),
+    /// and among all validators if everyone is already at the max.
+    /// This function ensures the following invariants:
+    /// 1. Total voting power of all validators sums to `TOTAL_VOTING_POWER`
+    /// 2. `active_validators` is sorted by voting power in descending order
+    /// 3. `active_validators` is sorted by stake in descending order
+    /// This function attempts to maintain the following invariants whenever possible:
+    /// 4. Each validator's voting power is <= `MAX_VOTING_POWER`
+    /// 5. If validator A and B have the same stake, they will have the same voting power
+    /// Invariant (4) and (5) should almost always hold for Sui in practice due to high validator count, and stakes that aren't exactly
+    /// equal, but in theory/in tests these can be violated due to:
+    /// - a staking distribution like [1, 1, 1, 1] (will violate (4))
+    /// - a staking distribution where at least one validator has `MAX_VOTING_POWER`, and there is a remainder after redistributing the
+    ///   leftovers due to odd numbers. In this case, a single validator will have 1 more than its proportionsal share
+    public fun update(active_validators: &mut vector<Validator>): u64 {
```

Comment:
It might be more clear if we separate the complex voting power calculation logic and the actual updating of each validator's voting power in the `&mut vector<Validator>`. The calculation function really just needs to take in a vector of (validator_index, stake) and outputs a vector of (validator_index, voting_power). The update function simply updates  according to the output vector.

Reply:
> I feel a bit nervous passing in a mut vec here and then change their ordering. Are we sure that there will be no more index reference to the list at the point of calling this function?
I will refactor to make sure this happens after any indexing (was already almost the case) + add a comment that makes it clear that this should always be the case.

> It might be more clear if we separate the complex voting power calculation logic and the actual updating of each validator's voting power i

I think the indirection via these indexes is actually a bit harder to read/write/reason about--it introduces the possibility of invalid indexes or index mixup, which is very hard to do in a direct loop that reads/writes/updates the stake and voting power of each validator in a single iter.